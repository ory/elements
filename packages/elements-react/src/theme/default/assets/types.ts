// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export type SVGIcon = React.FunctionComponent<
  React.ComponentProps<"svg"> & { size?: number }
>
