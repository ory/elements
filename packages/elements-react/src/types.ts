// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import {
  UiNode,
  UiNodeAnchorAttributes,
  UiNodeImageAttributes,
  UiNodeInputAttributes,
  UiNodeTextAttributes,
} from "@ory/client-fetch"
import {
  ComponentPropsWithoutRef,
  FormEventHandler,
  MouseEventHandler,
  PropsWithChildren,
} from "react"
import { ButtonVariants } from "./theme/default/components/form/button"

export type OryNodeButtonProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
} & Omit<ComponentPropsWithoutRef<"button">, "children"> &
  ButtonVariants

export type OryCurrentIdentifierProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
  onClick?: () => void
  href?: string
} & Omit<ComponentPropsWithoutRef<"button">, "children" | "onClick">

export type OryNodeAnchorProps = {
  attributes: UiNodeAnchorAttributes
  node: UiNode
} & Omit<ComponentPropsWithoutRef<"a">, "children">

export type OryNodeLabelProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
} & ComponentPropsWithoutRef<"label">

export type OryNodeTextProps = {
  attributes: UiNodeTextAttributes
  node: UiNode
}

export type OryCardLogoProps = Record<string, never>

/**
 * Props for the AuthMethodListItem component. This component is used
 * to render a single auth method in the AuthMethodList component.
 */
export type OryCardAuthMethodListItemProps = {
  onClick: () => void
  group: string
}

export type OryNodeImageProps = {
  attributes: UiNodeImageAttributes
  node: UiNode
}

/**
 * A generic type for the form values.
 */
export type FormValues = Record<string, string | boolean | number | undefined>

export type OryFormRootProps = ComponentPropsWithoutRef<"form"> & {
  onSubmit: FormEventHandler<HTMLFormElement>
}

export type OryNodeInputProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
  onClick?: MouseEventHandler
}

export type OryFormSectionContentProps = PropsWithChildren<{
  title?: string
  description?: string
}>

export type OryFormSectionFooterProps = PropsWithChildren
