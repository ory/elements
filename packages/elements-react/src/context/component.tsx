import { UiNode } from "@ory/client-fetch"
import { PropsWithChildren, createContext, useContext } from "react"
import { OryFlowComponents } from "../types"

type ComponentContextValue = {
  components: OryFlowComponents
  nodeSorter: (a: UiNode, b: UiNode, ctx: { flowType: string }) => number
}

const ComponentContext = createContext<ComponentContextValue>({
  components: null!, // fine because we throw an error if it's not provided
  nodeSorter: () => 0,
})

export function useComponents() {
  const ctx = useContext(ComponentContext)
  if (!ctx) {
    throw new Error("useComponents must be used within a ComponentProvider")
  }
  return ctx.components
}

export function useNodeSorter() {
  const ctx = useContext(ComponentContext)
  if (!ctx) {
    throw new Error("useComponents must be used within a ComponentProvider")
  }
  return ctx.nodeSorter
}

const defaultGroupOrder = [
  "oidc",
  "identifier_first",
  "default",
  "profile",
  "password",
  "passkey",
  "code",
  "webauthn",
]

function defaultNodeSorter(
  a: UiNode,
  b: UiNode,
  // ctx: { flowType: string },
): number {
  const aGroupWeight = defaultGroupOrder.indexOf(a.group) ?? 999
  const bGroupWeight = defaultGroupOrder.indexOf(b.group) ?? 999

  return aGroupWeight - bGroupWeight
}

type ComponentProviderProps = {
  components: OryFlowComponents
  nodeSorter?: (a: UiNode, b: UiNode, ctx: { flowType: string }) => number
}

export function OryComponentProvider({
  children,
  components,
  nodeSorter = defaultNodeSorter,
}: PropsWithChildren<ComponentProviderProps>) {
  return (
    <ComponentContext.Provider
      value={{
        components,
        nodeSorter,
      }}
    >
      {children}
    </ComponentContext.Provider>
  )
}
