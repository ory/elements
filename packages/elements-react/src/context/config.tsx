// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { createContext, PropsWithChildren, useContext, useRef } from "react"
import { isProduction } from "../client/config"
import { OryClientConfiguration } from "../util"
import { frontendClient } from "../util/client"
import {
  AccountExperienceConfiguration,
  ConfigurationParameters,
  FrontendApi,
} from "@ory/client-fetch"

export type OryElementsConfiguration = {
  sdk: OrySDK
  project: AccountExperienceConfiguration
}

const defaultProject: AccountExperienceConfiguration = {
  name: "Ory",
  registration_enabled: true,
  verification_enabled: true,
  recovery_enabled: true,
  recovery_ui_url: "/ui/recovery",
  registration_ui_url: "/ui/registration",
  verification_ui_url: "/ui/verification",
  login_ui_url: "/ui/login",
  settings_ui_url: "/ui/settings",
  default_redirect_url: "/ui/welcome",
  error_ui_url: "/ui/error",
  default_locale: "en",
  locale_behavior: "force_default",
}

export function useOryConfiguration(): OryElementsConfiguration {
  const configCtx = useContext(OryConfigurationContext)
  return {
    sdk: {
      ...configCtx.sdk,
      frontend: frontendClient(configCtx.sdk.url, configCtx.sdk.options),
    },
    project: {
      ...configCtx.project,
    },
  }
}

export type OrySDK = SDKConfig & {
  /**
   * The frontend client for the Ory SDK.
   * This client is used to interact with the Ory SDK and should be used to make API calls.
   */
  frontend: FrontendApi
}

type SDKConfig = {
  /**
   * The URL of the Ory SDK.
   * This URL is used to connect to the Ory SDK and should be set to the base URL of your Ory instance.
   */
  url: string
  options?: Partial<ConfigurationParameters>
}

type OryElementsConfigContextType = {
  sdk: SDKConfig
  project: AccountExperienceConfiguration
}

const OryConfigurationContext = createContext<OryElementsConfigContextType>({
  sdk: computeSdkConfig({}),
  project: defaultProject,
})

export function OryConfigurationProvider({
  children,
  sdk: initialConfig,
  project,
}: PropsWithChildren<{
  sdk?: Partial<OryClientConfiguration["sdk"]>
  project?: Partial<AccountExperienceConfiguration>
}>) {
  const configRef = useRef({
    sdk: computeSdkConfig(initialConfig),
    project: {
      ...defaultProject,
      ...project,
    },
  })

  return (
    <OryConfigurationContext.Provider value={configRef.current}>
      {children}
    </OryConfigurationContext.Provider>
  )
}

function computeSdkConfig(config?: OryClientConfiguration["sdk"]): SDKConfig {
  if (config?.url && typeof config.url === "string") {
    return {
      url: config.url.replace(/\/$/, ""),
      options: config.options || {},
    }
  }

  return {
    url: getSDKUrl(),
    options: config?.options || {},
  }
}

function getSDKUrl() {
  if (
    typeof process !== "undefined" &&
    process.versions &&
    process.versions.node
  ) {
    // process is available, let's try some environment variables
    if (isProduction()) {
      const sdkUrl =
        process.env["NEXT_PUBLIC_ORY_SDK_URL"] ?? process.env["ORY_SDK_URL"]
      if (!sdkUrl) {
        throw new Error(
          "Unable to determine SDK URL. Please set NEXT_PUBLIC_ORY_SDK_URL and/or ORY_SDK_URL in production environments.",
        )
      }
      return sdkUrl.replace(/\/$/, "")
    } else {
      if (process.env["__NEXT_PRIVATE_ORIGIN"]) {
        return process.env["__NEXT_PRIVATE_ORIGIN"].replace(/\/$/, "")
      } else if (process.env["VERCEL_URL"]) {
        return `https://${process.env["VERCEL_URL"]}`.replace(/\/$/, "")
      }
    }
  }

  if (typeof window !== "undefined") {
    // we are in the browser

    // Try to use window location
    return window.location.origin
  }
  // We aren't in node, and we don't have a window location.
  // This is probably a test environment, so we can't guess the SDK URL.

  throw new Error(
    "Unable to determine SDK URL. Please set NEXT_PUBLIC_ORY_SDK_URL and/or ORY_SDK_URL or supply the sdk.url parameter in the Ory configuration.",
  )
}
