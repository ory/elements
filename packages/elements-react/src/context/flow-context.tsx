// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import {
  Dispatch,
  PropsWithChildren,
  createContext,
  useContext,
  useState,
} from "react"
import { OryFlowContainer } from "../util/flowContainer"

/**
 * Returns an object that contains the current flow and the flow type, as well as the configuration.
 *
 * @returns The current flow container
 */
export function useOryFlow() {
  const ctx = useContext(OryFlowContext)
  if (!ctx) {
    throw new Error("useOryFlow must be used within a OryFlowProvider")
  }

  return ctx
}

/**
 * Function to set the flow container.
 */
export type FlowContainerSetter = Dispatch<Partial<OryFlowContainer>>

/**
 * The return value of the OryFlowContext.
 */
export type FlowContextValue = OryFlowContainer & {
  /**
   * Function to set the flow container.
   */
  setFlowContainer: FlowContainerSetter
}

// This is fine, because we don't export the context itself and guard from it being null in useOryFlow
const OryFlowContext = createContext<FlowContextValue>(null!)

export type OryFlowProviderProps = PropsWithChildren<OryFlowContainer>

export function OryFlowProvider({
  children,
  ...container
}: OryFlowProviderProps) {
  const [flowContainer, setFlowContainer] = useState(container)

  return (
    <OryFlowContext.Provider
      value={
        {
          ...flowContainer,
          setFlowContainer: (updatedContainer) => {
            setFlowContainer(
              (container) =>
                ({
                  ...container,
                  ...updatedContainer,
                }) as OryFlowContainer,
            )
          },
        } as FlowContextValue
      }
    >
      {children}
    </OryFlowContext.Provider>
  )
}
