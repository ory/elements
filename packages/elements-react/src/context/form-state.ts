// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { FlowType, UiNode, UiNodeGroupEnum } from "@ory/client-fetch"
import { useReducer } from "react"
import { isChoosingMethod } from "../components/card/card-two-step.utils"
import { OryFlowContainer } from "../util"
import { nodesToAuthMethodGroups } from "../util/ui"

export type FormState =
  | { current: "provide_identifier" }
  | { current: "select_method" }
  | { current: "method_active"; method: UiNodeGroupEnum }
  | { current: "success_screen" }
  | { current: "settings" }

export type FormStateAction =
  | {
      type: "action_flow_update"
      flow: OryFlowContainer
    }
  | {
      type: "action_select_method"
      method: UiNodeGroupEnum
    }

function findMethodWithMessage(nodes?: UiNode[]) {
  return nodes
    ?.filter((n) => !["default", "identifier_first"].includes(n.group))
    ?.find((node) => node.messages?.length > 0)
}

function parseStateFromFlow(flow: OryFlowContainer): FormState {
  switch (flow.flowType) {
    case FlowType.Registration:
    case FlowType.Login: {
      const methodWithMessage = findMethodWithMessage(flow.flow.ui.nodes)
      if (flow.flow.active == "link_recovery") {
        return { current: "method_active", method: "link" }
      } else if (flow.flow.active == "code_recovery") {
        return { current: "method_active", method: "code" }
      } else if (methodWithMessage) {
        return { current: "method_active", method: methodWithMessage.group }
      } else if (
        flow.flow.active &&
        !["default", "identifier_first", "oidc"].includes(flow.flow.active)
      ) {
        return { current: "method_active", method: flow.flow.active }
      } else if (isChoosingMethod(flow.flow.ui.nodes)) {
        // Login has a special case where we only have one method. Here, we
        // do not want to display the chooser.
        const authMethods = nodesToAuthMethodGroups(flow.flow.ui.nodes)
        if (authMethods.length === 1) {
          return { current: "method_active", method: authMethods[0] }
        }
        return { current: "select_method" }
      } else if (flow.flow.ui.messages?.some((m) => m.id === 1010016)) {
        // Account linking edge case
        return { current: "select_method" }
      }
      return { current: "provide_identifier" }
    }
    case FlowType.Recovery:
    case FlowType.Verification:
      // The API does not provide types for the active field of the recovery flow
      // TODO: Add types for the recovery flow in Kratos
      if (flow.flow.active === "code" || flow.flow.active === "link") {
        if (flow.flow.state === "choose_method") {
          return { current: "provide_identifier" }
        }
        return { current: "method_active", method: flow.flow.active }
      }
      break
    case FlowType.Settings:
      return { current: "settings" }
  }
  console.warn(
    `[Ory/Elements React] Encountered an unknown form state on ${flow.flowType} flow with ID ${flow.flow.id}`,
  )
  throw new Error("Unknown form state")
}

export function formStateReducer(
  state: FormState,
  action: FormStateAction,
): FormState {
  switch (action.type) {
    case "action_flow_update":
      return parseStateFromFlow(action.flow)
    case "action_select_method":
      return { current: "method_active", method: action.method }
  }
  return state
}

export function useFormStateReducer(flow: OryFlowContainer) {
  return useReducer(formStateReducer, parseStateFromFlow(flow))
}
