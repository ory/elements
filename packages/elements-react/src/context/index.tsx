// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export { useComponents, useNodeSorter } from "./component"
export {
  useOryFlow,
  type FlowContextValue,
  type FlowContainerSetter,
} from "./flow-context"
export * from "./provider"
