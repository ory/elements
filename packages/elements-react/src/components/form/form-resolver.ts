// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { useOryFlow } from "../../context"
import { FormValues } from "../../types"

function isCodeResendRequest(data: FormValues) {
  return data.email ?? data.resend
}

/**
 * Creates a resolver for the Ory form
 *
 * The resolver does form validation for missing fields in the form.
 *
 * @returns a react-hook-form resolver for the Ory form
 */
export function useOryFormResolver() {
  const flowContainer = useOryFlow()

  return (data: FormValues) => {
    if (flowContainer.formState.current === "method_active") {
      if (data.method === "code" && !data.code && !isCodeResendRequest(data)) {
        return {
          values: data,
          errors: {
            code: {
              id: 4000002,
              context: {
                property: "code",
              },
              type: "error",
              text: "Property code is missing",
            },
          },
        }
      }
    }
    return {
      values: data,
      errors: {},
    }
  }
}
