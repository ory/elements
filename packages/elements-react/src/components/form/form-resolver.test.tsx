// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { renderHook } from "@testing-library/react"
import { FormValues } from "../../types"
import { useOryFormResolver } from "./form-resolver"
import { OryFlowProvider } from "../../context/flow-context"
import { defaultConfiguration } from "../../tests/jest/test-utils"
import { PropsWithChildren } from "react"
import { FlowType, LoginFlow } from "@ory/client-fetch"

const testCases = [
  {
    method: "code",
    code: "",
    email: "",
  },
  {
    method: "code",
  },
  {
    method: "code",
    code: "123456",
    email: "",
  },
  {
    method: "code",
    code: "123456",
    email: "some@example.com",
  },
  {
    method: "password",
  },
] satisfies FormValues[]

const wrapper = ({ children }: PropsWithChildren) => (
  <OryFlowProvider
    config={defaultConfiguration}
    flow={
      {
        active: "code",
        ui: { nodes: [], action: "", method: "" },
      } as unknown as LoginFlow // Fine, we're just testing the resolver
    }
    flowType={FlowType.Login}
  >
    {children}
  </OryFlowProvider>
)

for (const testCase of testCases) {
  test("case=" + JSON.stringify(testCase), () => {
    const formResolver = renderHook(() => useOryFormResolver(), {
      wrapper,
    })
    expect(formResolver.result.current(testCase)).toMatchSnapshot()
  })
}
