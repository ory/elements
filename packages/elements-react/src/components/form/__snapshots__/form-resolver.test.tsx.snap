// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`case={"method":"code","code":"","email":""} 1`] = `
{
  "errors": {
    "code": {
      "context": {
        "property": "code",
      },
      "id": 4000002,
      "text": "Property code is missing",
      "type": "error",
    },
  },
  "values": {
    "code": "",
    "email": "",
    "method": "code",
  },
}
`;

exports[`case={"method":"code","code":"123456","email":""} 1`] = `
{
  "errors": {},
  "values": {
    "code": "123456",
    "email": "",
    "method": "code",
  },
}
`;

exports[`case={"method":"code","code":"123456","email":"some@example.com"} 1`] = `
{
  "errors": {},
  "values": {
    "code": "123456",
    "email": "some@example.com",
    "method": "code",
  },
}
`;

exports[`case={"method":"code"} 1`] = `
{
  "errors": {
    "code": {
      "context": {
        "property": "code",
      },
      "id": 4000002,
      "text": "Property code is missing",
      "type": "error",
    },
  },
  "values": {
    "method": "code",
  },
}
`;

exports[`case={"method":"password"} 1`] = `
{
  "errors": {},
  "values": {
    "method": "password",
  },
}
`;
